packages/stellar-axelar-std-derive/src/storage.rs

```diff
@@ -0,0 +1,400 @@
+        }
+
+        impl #name {
+            #(#storage_fns)*
+        }
+
+        #(#public_fns)*
+    };
+
+    output
+}
+
+/// Transforms a contractstorage enum variant with named fields into a storage key map (tuple variant),
+/// or a single storage key for a unit variant.
+///
+/// The Unit variant must be captured here to avoid suffixing non-map variants with "{}".
+///
+/// # Example
+/// ```rust,ignore
+/// /* Original */
+/// #[contractstorage]
+/// enum DataKey {
+///     #[instance]
+///     #[value(Address)]
+///     Gateway, // Unit variant (would need `Gateway {},` otherwise)
+///
+///     #[temporary]
+///     #[value(Address)]
+///     Users { user: Address }, // Named variant (one or more fields)
+/// }
+/// ```
+///
+/// /* Generated */
+/// #[contracttype]
+/// pub enum DataKey {
+///     Gateway, // Unit variant (storage key)
+///     Users(Address), // Tuple variant (storage key map)
+/// }
+/// ```
+fn transform_variant(variant: &Variant) -> TokenStream {
+    let variant_name = &variant.ident;
+
+    match &variant.fields {
+        Fields::Unit => {
+            quote! {
+                #variant_name
+            }
+        }
+        Fields::Named(FieldsNamed { named, .. }) => {
+            let types = named.iter().map(|f| &f.ty);
+            quote! {
+                #variant_name(#(#types),*)
+            }
+        }
+        _ => panic!("Only unit variants or named fields are supported in storage enums."),
+    }
+}
+
+/// Returns the storage type of a storage enum variant.
+fn storage_type(attrs: &[Attribute]) -> StorageType {
+    attrs.iter().flat_map(|attr| match attr {
+        _ if attr.path().is_ident("instance") => Some(StorageType::Instance),
+        _ if attr.path().is_ident("persistent") => Some(StorageType::Persistent),
+        _ if attr.path().is_ident("temporary") => Some(StorageType::Temporary),
+        _ => None})
+    .exactly_one()
+    .expect("Storage type must be specified exactly once as 'instance', 'persistent', or 'temporary'.")
+}
+
+/// Returns the value type of a storage enum variant.
+fn value_type(attrs: &[Attribute]) -> Type {
+    attrs
+        .iter()
+        .flat_map(|attr| {
+            if attr.path().is_ident("value") {
+                if let Meta::List(list) = &attr.meta {
+                    Some(
+                        list.parse_args::<Type>()
+                            .expect("Failed to parse value type."),
+                    )
+                } else {
+                    panic!("Value attribute must contain a type parameter: #[value(Type)]");
+                }
+            } else {
+                None
+            }
+        })
+        .exactly_one()
+        .expect("Missing required #[value(Type)] attribute.")
+}
+
+/// Generates the storage getter, setter, and deleter functions for a storage enum variant.
+fn storage_fns(
+    enum_name: &Ident,
+    variant: &Variant,
+    storage_attrs: &StorageAttributes,
```

Comment:
you can destructure in the arg itself, `StorageAttributes { storage_types, value_types}: StorageAttributes`